src/parse.c

```diff
@@ -10467,10 +10469,145 @@ static tree_t p_psl_directive(void)
    return t;
 }
 
-static tree_t p_psl_clock_declaration(void)
+static type_t p_psl_subtype_indication(void)
+{
+   BEGIN("PSL subtype indication");
+
+   // Handle ambiguity in "p_subtype_indication". Two consecutive IDs
+   // may indicate resolution function, or only "type_mark" followed by
+   // actual parameter name.
+   token_t third = peek_nth(3);
+   if (peek() == tID && peek_nth(1) == tID &&
+       (third == tSEMI || third == tCOMMA || third == tRPAREN))
+      return p_type_mark();
+   else
+      return p_subtype_indication();
+}
+
+static type_t p_psl_param_spec(psl_node_t node, psl_type_t *psl_type, class_t *class)
+{
+   //    const
+   //    | [const | mutable] Value_Parameter
+   //    | sequence
+   //    | property
+   //
+   //    Value_Parameter ::=
+   //       HDL_Type
+   //     | PSL_Type_Class
+   //
+   //    HDL_Type ::=
+   //       hdltype HDL_VARIABLE_TYPE
+   //
+   //    PSL_Type_Class ::=
+   //       boolean | bit | bitvector | numeric | string
+
+   BEGIN("PSL Parameter specification");
+
+   *class = C_MUTABLE;
+   token_t ttok;
+
+   switch (peek()) {
+   case tPROPERTY:
+      consume(tPROPERTY);
+      *psl_type = PSL_TYPE_PROPERTY;
+      break;
+
+   case tSEQUENCE:
+      consume(tSEQUENCE);
+      *psl_type = PSL_TYPE_SEQUENCE;
+      break;
+
+   case tCONST:
+      *class = C_CONSTANT;
+      // Handle PSL 1.1 "const" only
+      if (peek_nth(2) == tID) {
+         consume(tCONST);
+         *psl_type = PSL_TYPE_NUMERIC;
+         return std_type(NULL, STD_INTEGER);
+      }
+   case tMUTABLE:
+      one_of(tCONST, tMUTABLE);
+   case tHDLTYPE:
+   case tBOOLEAN:
+   case tBIT:
+   case tBITVECTOR:
+   case tNUMERIC:
+   case tSTRINGK:
+      ttok = one_of(tHDLTYPE, tBOOLEAN, tBIT, tBITVECTOR, tNUMERIC, tSTRINGK);
+      switch (ttok) {
+      case tHDLTYPE:
+         *psl_type = PSL_TYPE_HDLTYPE;
+         scan_as_vhdl();
+         type_t t = p_psl_subtype_indication();
+         scan_as_psl();
+         return t;
+      case tBOOLEAN:
+         *psl_type = PSL_TYPE_BOOLEAN;
+         return std_type(NULL, STD_BOOLEAN);
+      case tBIT:
+         *psl_type = PSL_TYPE_BIT;
+         return std_type(NULL, STD_BIT);
+         break;
+      case tBITVECTOR:
+         *psl_type = PSL_TYPE_BITVECTOR;
+         return std_type(NULL, STD_BIT_VECTOR);
+         break;
+      case tNUMERIC:
+         *psl_type = PSL_TYPE_NUMERIC;
+         return std_type(NULL, STD_INTEGER);
+         break;
+      case tSTRINGK:
+         *psl_type = PSL_TYPE_STRING;
+         return std_type(NULL, STD_STRING);
+         break;
+      default:
+         one_of(tHDLTYPE, tBOOLEAN, tBIT, tBITVECTOR, tNUMERIC, tSTRINGK);
```

Comment:
This is covered by the previous call to `one_of` so can just return NULL here or let it fall through to the return below.

Does returning NULL cause any problems later if the input is invalid? Normally I return a dummy type `T_NONE` in cases where there was an error in the type. E.g. `type_new(T_NONE)`. That prevents cascading errors (it compares equal to every other type) and avoid having to check `tree_has_type()` all the time.